Feat: Upstream Oauth plugin (#7814)
Kong/docs.konghq.com: app/_hub/kong-inc/upstream-oauth/overview/_index.md

---
nav_title: Overview
title: Overview
---

The Upstream OAuth plugin allows {{site.base_gateway}} to support OAuth flows between Kong and the upstream API.

The plugin supports storing tokens issued by the IdP in different backend formats.

## How it works

The upstream OAuth2 credential flow works similarly to the [client credentials grant](/hub/kong-inc/openid-connect/how-to/authentication/client-credentials/) used by the OpenID Connect plugin. If a cached access token isn't found, Kong issues a request to the IdP token endpoint to obtain a new token, which is cached, and then passed to the upstream API via a configurable header (`Authorization` by default).

<!--vale off-->

{% mermaid %}
sequenceDiagram
    autonumber
    participant client as Client <br>(e.g. mobile app)
    participant kong as API Gateway <br>(Kong)

>     participant idp as IDP <br>(e.g. Keycloak)

Reviewer: ```suggestion
    participant idp as IdP <br>(e.g. Keycloak)
```

    participant api as 3rd Party API
    activate client
    activate kong
    client->>kong: request to Kong
    deactivate client
    activate idp
    kong->>idp: request access token <br>from IdP using <br>client ID and client secret (if IdP auth is set)
    deactivate kong
    idp->>idp: authenticate client
    activate kong
    idp->>kong: return access token
    deactivate idp
    activate api

>     kong->>api: request with access token in <br>Authorization header

Reviewer: ```suggestion
    kong->>api: request with access token in <br><br>Authorization header
```
I don't know if this will work but on this screen it looks like this: 
![image](https://github.com/user-attachments/assets/62ef9e61-ebc6-4ec2-a998-97f0e5e84d76)

    deactivate kong
    activate kong
    api->>kong: response
    deactivate api
    activate client
    kong->>client: response
    deactivate client
    deactivate kong
{% endmermaid %}

<!--vale on-->

## Authentication methods

This plugin supports the following authentication methods:

* [`client_secret_basic`](/hub/kong-inc/upstream-oauth/configuration/#config-client_secret_basic): Send `client_id` and `client_secret` in an `Authorization: Basic` header
* [`client_secret_post`](/hub/kong-inc/upstream-oauth/configuration/#config-client_secret_basic): Send `client_id` and `client_secret` as part of the body
* [`client_secret_jwt`](/hub/kong-inc/upstream-oauth/configuration/#config-client_secret_basic): Send a JWT signed with the `client_secret` using the client assertion as part of the body

## Caching

The Upstream OAuth plugin caches tokens returned by the IdP.  

> Cache entries expire based on the `expires_in` indicated by the IdP in the response to the token endpoint.

Reviewer: ```suggestion
Cached entries expire based on the `expires_in` indicated by the IdP in the response to the token endpoint.
```
cached because the entries are already cached?

Tokens are cached using a hash of all values configured under the [`config.oauth`](/hub/kong-inc/upstream-oauth/configuration/#config-oauth) key.
This means that if two instances of the plugin (for example, configured on different routes and services) use identical values under the `config.oauth` section,
then these will share cached tokens.

### Caching strategies

The plugin supports the following caching [strategies](/hub/kong-inc/upstream-oauth/configuration/#config-stategy):

> * `memory`: A locally stored `lua_shared_dict`. Note that the default dictionary, `kong_db_cache`, is also used by other plugins and elements of Kong to store unrelated database cache entities.

Reviewer: ```suggestion
* `memory`: A locally stored `lua_shared_dict`. The default dictionary, `kong_db_cache`, is also used by other plugins and {{site.base_gateway}} elements to store unrelated database cache entities.
```

---
nav_title: Overview
title: Overview
---

> The Upstream OAuth plugin lets you manage OAuth flows between {{site.base_gateway}} and an upstream service.

Reviewer 2: ```suggestion
The Upstream OAuth plugin allows you to manage OAuth flows between {{site.base_gateway}} and an upstream service.
```
This could be me not understanding auth plugins well, but why would I use this one instead of a different auth plugin? Do those other ones not allow you to manage OAuth with an upstream service? How would I know I need this one and not the other ones?

Author: It's not you, this is poorly phrased. I'll rephrase.

> The Upstream OAuth plugin lets you manage OAuth flows between {{site.base_gateway}} and an upstream service.

Author: ```suggestion
The Upstream OAuth plugin allows {{site.base_gateway}} to consume third-party or external APIs to initiate OAuth flows, which lets you use {{site.base_gateway}} in egress workflows.
```

> The Upstream OAuth plugin allows {{site.base_gateway}} to consume third-party or external APIs to initiate OAuth flows, which lets you use {{site.base_gateway}} in egress workflows.

Reviewer 3: ```suggestion
The Upstream OAuth plugin allows {{site.base_gateway}} to support OAuth flows between Kong and the upstream API.
```

> The plugin is designed to support storing tokens issued by the IdP in different backend formats.

Reviewer 2: ```suggestion
The plugin supports storing tokens issued by the IdP in different backend formats.
```

## How it works

> The upstream OAuth2 credential flow works similarly to the [client credentials grant](/hub/kong-inc/openid-connect/how-to/authentication/client-credentials/) used by the OpenID Connect plugin. However, instead of Kong verifying the access token, the verification is passed along to the upstream API, which then validates the token against the IdP.

Reviewer 3: ```suggestion
The upstream OAuth2 credential flow works similarly to the [client credentials grant](/hub/kong-inc/openid-connect/how-to/authentication/client-credentials/) used by the OpenID Connect plugin. If a cached access token isn't found, Kong issues a request to the IdP token endpoint to obtain a new token, which is cached, and then passed to the upstream API via a configurable header (`Authorization` by default).
```

<!--vale off-->

{% mermaid %}
sequenceDiagram
    autonumber
    participant client as Client <br>(e.g. mobile app)
    participant kong as API Gateway <br>(Kong)
    participant api as 3rd Party API
    participant idp as IDP <br>(e.g. Keycloak)
    activate client
    activate kong
    client->>kong: request to Kong<br>with any supported<br> authentication method
    deactivate client
    kong->>kong: load authentication<br> credentials
    activate idp
    kong->>idp: request access token <br>from IdP using <br>client ID and client secret
    deactivate kong
    idp->>idp: authenticate client
    activate kong
    idp->>kong: return access token
    deactivate idp
    activate api
    kong->>api: request with access token in <br>Authorization header
    deactivate kong
    activate idp
    api->>idp: validates <br> access token via IdP
    idp->>api: grants access
    deactivate idp
    activate kong
    api->>kong: response
    deactivate api
    activate client
    kong->>client: response
    deactivate client
    deactivate kong

> {% endmermaid %}

Author: {% mermaid %}
sequenceDiagram
    autonumber
    participant client as Client <br>(e.g. mobile app)
    participant kong as API Gateway <br>(Kong)
    participant idp as IDP <br>(e.g. Keycloak)
    participant api as 3rd Party API
    activate client
    activate kong
    client->>kong: request to Kong
    deactivate client
    activate idp
    kong->>idp: request access token <br>from IdP using <br>client ID and client secret (if IdP auth is set)
    deactivate kong
    idp->>idp: authenticate client
    activate kong
    idp->>kong: return access token
    deactivate idp
    activate api
    kong->>api: request with access token in <br>Authorization header
    deactivate kong
    activate kong
    api->>kong: response
    deactivate api
    activate client
    kong->>client: response
    deactivate client
    deactivate kong
{% endmermaid %}
```

> {% mermaid %}
> sequenceDiagram
>     autonumber
>     participant client as Client <br>(e.g. mobile app)
>     participant kong as API Gateway <br>(Kong)
>     participant api as 3rd Party API
>     participant idp as IDP <br>(e.g. Keycloak)
>     activate client
>     activate kong
>     client->>kong: request to Kong<br>with any supported<br> authentication method
>     deactivate client
>     kong->>kong: load authentication<br> credentials
>     activate idp
>     kong->>idp: request access token <br>from IdP using <br>client ID and client secret
>     deactivate kong
>     idp->>idp: authenticate client
>     activate kong
>     idp->>kong: return access token
>     deactivate idp
>     activate api
>     kong->>api: request with access token in <br>Authorization header
>     deactivate kong
>     activate idp
>     api->>idp: validates <br> access token via IdP
>     idp->>api: grants access
>     deactivate idp
>     activate kong
>     api->>kong: response
>     deactivate api
>     activate client
>     kong->>client: response
>     deactivate client
>     deactivate kong
> {% endmermaid %}

Author: ```suggestion
{% mermaid %}
sequenceDiagram
    autonumber
    participant client as Client <br>(e.g. mobile app)
    participant kong as API Gateway <br>(Kong)
    participant idp as IDP <br>(e.g. Keycloak)
    participant api as 3rd Party API
    activate client
    activate kong
    client->>kong: request to Kong
    deactivate client
    activate idp
    kong->>idp: request access token <br>from IdP using <br>client ID and client secret (if IdP auth is set)
    deactivate kong
    idp->>idp: authenticate client
    activate kong
    idp->>kong: return access token
    deactivate idp
    activate api
    kong->>api: request with access token in <br>Authorization header
    deactivate kong
    activate kong
    api->>kong: response
    deactivate api
    activate client
    kong->>client: response
    deactivate client
    deactivate kong
{% endmermaid %}
```

<!--vale on-->

## Authentication methods

This plugin supports the following authentication methods:

* [`client_secret_basic`](/hub/kong-inc/upstream-oauth/configuration/#config-client_secret_basic): Send `client_id` and `client_secret` in an `Authorization: Basic` header
* [`client_secret_post`](/hub/kong-inc/upstream-oauth/configuration/#config-client_secret_basic): Send `client_id` and `client_secret` as part of the body
* [`client_secret_jwt`](/hub/kong-inc/upstream-oauth/configuration/#config-client_secret_basic): Send a JWT signed with the `client_secret` using the client assertion as part of the body

## Caching

The Upstream OAuth plugin caches tokens returned by the IdP.  
Cache entries expire based on the `expires_in` indicated by the IdP in the response to the token endpoint.

Tokens are cached using a hash of all values configured under the [`config.oauth`](/hub/kong-inc/upstream-oauth/configuration/#config-oauth) key.
This means that if two instances of the plugin (for example, configured on different routes and services) use identical values under the `config.oauth` section,
then these will share cached tokens.

### Caching strategies

> The plugin supports the following caching [strategies](/hub/kong-inc/upstream-oauth/configuration/#config-stategy) are provided:

Reviewer 2: ```suggestion
The plugin supports the following caching [strategies](/hub/kong-inc/upstream-oauth/configuration/#config-stategy):
```

> * `memory`: A `lua_shared_dict`. Note that the default dictionary, `kong_db_cache`, is also used by other plugins and elements of Kong to store unrelated database cache entities.

Reviewer 2: Is this local storage? I think the note is implying to be careful how much you store here in case you run out of space?

> * `memory`: A `lua_shared_dict`. Note that the default dictionary, `kong_db_cache`, is also used by other plugins and elements of Kong to store unrelated database cache entities.

Author: ```suggestion
* `memory`: A locally stored `lua_shared_dict`. Note that the default dictionary, `kong_db_cache`, is also used by other plugins and elements of Kong to store unrelated database cache entities.
```
